resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -170,91 +206,294 @@ private[spark] class KubernetesClusterSchedulerBackend(
+                    runningExecutorPods(executorId).getMetadata.getName)
+                  .get()
+                val nodeName = runningExecutorPod.getSpec.getNodeName
+                val shufflePodIp = nodeCache.getShufflePodForExecutor(nodeName)
+                resolvedProperties = resolvedProperties ++ Seq(
+                  (SPARK_SHUFFLE_SERVICE_HOST.key, shufflePodIp))
+
+                val reply = SparkAppConfig(
+                  resolvedProperties,
+                  SparkEnv.get.securityManager.getIOEncryptionKey())
+                context.reply(reply)
+              }
+          }
+        }
+      }.orElse(super.receiveAndReply(context))
+    }
+  }
+
+  /*
+   * KubernetesAllocator class watches all the executor pods associated with
+   * this SparkJob and creates new executors when it is appropriate.
+   */
+  private[spark] class KubernetesAllocator(client: KubernetesClient,
+      interval: Int)
+      extends Logging {
+
+    private var scheduler: ScheduledExecutorService = _
+    private var watcher: Watch = _
+    private var readyExecutorPods = Set[String]()
+
+    def start(): Unit = {
+      scheduler = ThreadUtils
+        .newDaemonSingleThreadScheduledExecutor("kubernetes-pod-allocator")
+
+      watcher = client
+        .pods()
+        .withLabels(Map(SPARK_APP_ID_LABEL -> applicationId()).asJava)
+        .watch(new Watcher[Pod] {
+          override def eventReceived(action: Watcher.Action, p: Pod): Unit = {
+            action match {
+              case Action.DELETED =>
+                readyExecutorPods -= p.getMetadata.getName
+
+              case _ =>
+                if (Readiness.isReady(p)) {
+                  readyExecutorPods += p.getMetadata.getName
+                }
+            }
+          }
+          override def onClose(e: KubernetesClientException): Unit = {}
+        })
+
+      if (interval > 0) {
+        scheduler.scheduleWithFixedDelay(allocatorRunnable,
+          0,
+          interval,
+          TimeUnit.SECONDS)
+      }
+    }
+
+    def stop(): Unit = {
+      watcher.close()
+      scheduler.shutdown()
+    }
+
+    private def getAllocationSize(): Int = {
+      val allocationChunkSize = conf.get(KUBERNETES_DYNAMIC_ALLOCATION_SIZE)
+      math.min(executorLimitOption.get - runningExecutorPods.size, allocationChunkSize)
+    }
+
+    private val allocatorRunnable: Runnable = new Runnable {
+      override def run(): Unit = {
+        if (readyExecutorPods.size < totalExpectedExecutors.get * minRegisteredRatio) {
+          logDebug("Waiting for pending executors before scaling")
+          return
+        }
+
+        if (runningExecutorPods.size >= executorLimitOption.get) {
+          logDebug(
+            "Maximum allowed executor limit reached. Not scaling up further.")
+          return
+        }
+
+        EXECUTOR_MODIFICATION_LOCK.synchronized {
+          for (i <- 1 to math.min(
+            executorLimitOption.get - runningExecutorPods.size,
```

Comment:
We don't need a separate executor limit here, right? Can we not use the expected number of executors, and then block on the executors we've specifically requested in this round to be ready? I think it's less clear if we have to track more global state in having both an executor limit and an expected executor count.